javascript/ql/src/semmle/javascript/frameworks/Electron.qll

```diff
@@ -73,187 +73,109 @@ module Electron {
+      override predicate canReturnTo(EventEmitter::EventDispatch dispatch) {
+        dispatch.(IPCDispatch).getCalleeName() = "sendSync"
+      }
     }
 
     /**
-     * An asynchronous Electron IPC message sent from the main process via a `webContents` object.
+     * A dispatch of an IPC event.
+     * An IPC event is sent from the Renderer to the Main process.
+     * And a value can be returned through the `returnValue` property of the event (first parameter in the callback).
      */
-    class WebContentsMessage extends Message {
-      DataFlow::Node channel;
-
-      WebContentsMessage() {
-        exists(WebContents wc, DataFlow::MethodCallNode mc |
-          wc.flowsTo(mc.getReceiver()) and
-          this = mc.getArgument(1) and
-          channel = mc.getArgument(0) and
-          mc.getCalleeName() = "send"
+    class IPCDispatch extends EventEmitter::EventDispatch::Range, DataFlow::InvokeNode {
+      override Process emitter;
+
+      IPCDispatch() {
+        exists(string methodName | methodName = "sendSync" or methodName = "send" |
+          this = emitter.ref().getAMemberCall(methodName)
         )
       }
 
-      override Process getProcess() { result = Process::main() }
-
-      override string getChannelName() { result = channel.getStringValue() }
-    }
-
-    /**
-     * Holds if `pred` flows to `succ` via Electron IPC.
-     */
-    private predicate ipcFlowStep(DataFlow::Node pred, DataFlow::Node succ) {
-      // match a message sent from one process with a callback parameter in the other process
-      exists(Callback callback, Message msg |
-        callback.getChannelName() = msg.getChannelName() and
-        callback.getProcess() != msg.getProcess() and
-        pred = msg and
-        succ = callback.getMessage()
-      )
-      or
-      // match a synchronous reply sent from one process with a `sendSync` call in the other process
-      exists(SyncDirectMessage sendSync, SyncReplyMessage msg |
-        sendSync.getChannelName() = msg.getChannelName() and
-        sendSync.getProcess() != msg.getProcess() and
-        pred = msg and
-        succ = sendSync.getReply()
-      )
-    }
-
-    /**
-     * An additional flow step  via an Electron IPC message.
-     */
-    private class IPCAdditionalFlowStep extends DataFlow::AdditionalFlowStep {
-      IPCAdditionalFlowStep() { ipcFlowStep(this, _) }
+      /**
+       * Gets the `i`th dispatched argument to the event handler.
+       * The 0th parameter in the callback is a event generated by the IPC system,
+       * therefore these arguments start at 1.
+       */
+      override DataFlow::Node getSentItem(int i) {
+        i >= 1 and
+        result = getArgument(i)
+      }
 
-      override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
-        pred = this and
-        ipcFlowStep(pred, succ)
+      /**
+       * Gets a registration that this dispatch can send an event to. 
+       */
+      override IPCSendRegistration getAReceiver() {
```

Comment:
This is now no longer symmetric with `canReturnTo`: one is a predicate with result, the other a plain unary predicate. You might consider adjusting them to be more consistent.